Plugins/MeshPaint/src/MeshPaintComponent.cpp

```diff
@@ -60,94 +84,115 @@ void MeshPaintComponent::startPaint( bool on ) {
     {
         ro->getRenderTechnique()->setConfiguration(
             Ra::Engine::ShaderConfigurationFactory::getConfiguration( "Plain" ) );
-        ro->getMesh()->addData( Ra::Engine::Mesh::VERTEX_COLOR, m_paintColors );
     } else
     {
+        Ra::Engine::Mesh* mesh = dynamic_cast<Ra::Engine::Mesh*>( ro->getMesh().get() );
+        CORE_ASSERT( mesh != nullptr, "Cannot access Mesh from Ro containing a TriangleMesh" );
+        ON_DEBUG( auto colAttrib =
+                      mesh->getTriangleMesh().getAttribHandle<Ra::Core::Vector4>( colAttribName ) );
+        CORE_ASSERT( colAttrib == m_currentColorAttribHdl, "Inconsistent AttribHandle used" );
+        // Could also be accessed using
+        // auto triangleMesh = compMess->get<Ra::Core::Geometry::TriangleMesh>( getEntity(),
+        // m_dataId ); however here we skip the search in the component map
+        Ra::Core::Geometry::TriangleMesh& triangleMesh = mesh->getTriangleMesh();
+        triangleMesh.getAttrib( m_currentColorAttribHdl ).data() = m_baseColors;
+
         ro->getRenderTechnique()->setConfiguration( m_baseConfig );
-        ro->getMesh()->addData( Ra::Engine::Mesh::VERTEX_COLOR, m_baseColors );
     }
 }
 
 void MeshPaintComponent::paintMesh( const Ra::Engine::Renderer::PickingResult& picking,
                                     const Ra::Core::Utils::Color& color ) {
+    using Ra::Engine::Mesh;
     // check it's for us
     if ( *m_renderObjectReader() != picking.m_roIdx || picking.m_mode == Ra::Engine::Renderer::RO )
     {
         return;
     }
 
     auto ro = getRoMgr()->getRenderObject( *m_renderObjectReader() );
-    if ( ro->getMesh()->getRenderMode() == Ra::Engine::Mesh::RM_POINTS &&
-         ( picking.m_mode != Ra::Engine::Renderer::VERTEX &&
-           picking.m_mode != Ra::Engine::Renderer::C_VERTEX ) )
+    auto pickingRenderMode = ro->getMesh()->pickingRenderMode();
+
+    if ( pickingRenderMode == Ra::Engine::Displayable::NO_PICKING )
     {
         return;
     }
 
-    if ( ( ro->getMesh()->getRenderMode() == Ra::Engine::Mesh::RM_LINES ||
-           ro->getMesh()->getRenderMode() == Ra::Engine::Mesh::RM_LINE_LOOP ||
-           ro->getMesh()->getRenderMode() == Ra::Engine::Mesh::RM_LINE_STRIP ||
-           ro->getMesh()->getRenderMode() == Ra::Engine::Mesh::RM_LINES_ADJACENCY ||
-           ro->getMesh()->getRenderMode() == Ra::Engine::Mesh::RM_LINE_STRIP_ADJACENCY ||
-           ro->getMesh()->getRenderMode() == Ra::Engine::Mesh::RM_TRIANGLE_STRIP ||
-           ro->getMesh()->getRenderMode() == Ra::Engine::Mesh::RM_TRIANGLE_FAN ) )
```

Comment:
Because with this changes, this is the Mesh that takes care of computing the right render mode, see `Mesh::updatePickingRenderMode`. 
So the correct version should be:
```c++
    if ( pickingRenderMode == Ra::Engine::Displayable::NO_PICKING ||
         pickingRenderMode == Ra::Engine::Displayable::PKM_LINES ||   // not supported yet
         pickingRenderMode == Ra::Engine::Displayable::PKM_LINE_ADJ ) // not supported yet
    {
        return;
    }
    if ( pickingRenderMode == Ra::Engine::Displayable::PKM_POINTS &&
         ( picking.m_mode != Ra::Engine::Renderer::VERTEX &&
           picking.m_mode != Ra::Engine::Renderer::C_VERTEX ) )
    {
        return;
    }
```
Good for you ?

Reply:
This sounds good for lines and ponts, but what about `RM_TRIANGLE_STRIP` and `RM_TRIANGLE_FAN` modes? These are baked into `PKM_TRIANGLES`, we then have to also check these separately

Reply:
Fixed.